>异常：一种导致程序终端执行的程序流。就是一种非致命的错误，通过修改可以能够完成程序的执行。所以需要完善的异常机制，保证程序的执行。

>处理异常：可以使用的关键字 try, catch , finally这几个关键字实现异常的捕获处理。
```
try{
	//可能出现异常的语句
}
	catch（异常类型 异常对象）{//捕获范围小的异常放在前
}
	catch（异常类型 异常对象）{//捕获范围大的异常放在后
}.....
	finally{
	不管异常是否处理都要执行的语句；
}
```
>可以使用的形式 **try  catch**；**try catch finally**；**try  finally**；三种通常会使用printStackTrace()方法打印异常；finally通常作为异常处理的出口；

>异常处理流程：异常类型![](https://upload-images.jianshu.io/upload_images/14935748-10318c9a0665233d.jpg?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
Java代码异常处理：![](https://upload-images.jianshu.io/upload_images/14935748-e4d200ae8a37b119.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
如果异常捕获没有捕获到，JVM会做异常默认的处理。

>throws关键字用于抛出异常，**常常放在方法名后**。如果主方法抛出异常则会交给JVM处理。
>throw关键字，主要作用是产生一个异常并手工的抛出一个异常的实例化对象在**代码块中使用**

>异常处理的标准格式：常常用在资源访问

```
public static 类型 方法名 throws（参数...）{
	try{
		//可能产生异常的语句
	}
	finally{
		//必须执行的语句例如网络连接关闭
	}
}
```

>assert断言：确定代码执行到某行得到预期的结果,java中的断言主要是通过assert关键字来执行一个表达式判断进行
```
//如果表达式为false，则抛出java.lang.AssertionError异常
assert expression;

//如果表达式为false，则抛出java.lang.AssertionError异常并输出错误信息errormessage
assert expressiion : message;
```
```
public class AssertTest {

    public static void main(String[] args) {

        test(-1);
    }

    /**
     * 断言检查
     * @param a
     */
    public static void test(int a){
        assert a>0 : "输入小于或等于0";
        System.out.println("断言正确，继续执行");
    }
}
```
>执行时加入-ea
![](https://upload-images.jianshu.io/upload_images/14935748-674c4efee2c6ea27.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

